File: src/algorithms/maxFlow.ts
import { GraphData, AlgorithmExecutionResult, AlgorithmStep } from '../types/graph';

interface FlowEdge {
  from: string;
  to: string;
  capacity: number;
  flow: number;
  revIndex: number;
  originalEdgeId: string;
}

export function runDinicMaxEvacuationFlow(
  graph: GraphData,
  sourceNodeIds?: string[],
  sinkNodeIds?: string[]
): AlgorithmExecutionResult {
  const startTime = performance.now();
  const steps: AlgorithmStep[] = [];

  // If no source/sink provided, auto-detect sources as nodes with highest hazard risk and sinks as shelters/hospitals
  const sources = sourceNodeIds && sourceNodeIds.length > 0
    ? sourceNodeIds
    : Object.values(graph.nodes).filter(n => n.hazardRisk > 0.4 || n.type === 'residential').map(n => n.id);

  const sinks = sinkNodeIds && sinkNodeIds.length > 0
    ? sinkNodeIds
    : Object.values(graph.nodes).filter(n => n.type === 'shelter' || n.type === 'hospital').map(n => n.id);

  const superSource = '__SUPER_SOURCE__';
  const superSink = '__SUPER_SINK__';

  // Build adjacency list for Dinic
  const adj: Record<string, FlowEdge[]> = {};
  const allNodes = [superSource, superSink, ...Object.keys(graph.nodes)];
  for (const id of allNodes) {
    adj[id] = [];
  }

  function addEdge(from: string, to: string, cap: number, edgeId: string) {
    const forward: FlowEdge = { from, to, capacity: cap, flow: 0, revIndex: adj[to].length, originalEdgeId: edgeId };
    const backward: FlowEdge = { from: to, to: from, capacity: 0, flow: 0, revIndex: adj[from].length, originalEdgeId: edgeId };
    adj[from].push(forward);
    adj[to].push(backward);
  }

  // Connect super-source to all disaster sources with their population demands
  for (const sId of sources) {
    const node = graph.nodes[sId];
    const pop = node ? Math.max(node.population, 500) : 1000;
    addEdge(superSource, sId, pop, `src_${sId}`);
  }

  // Connect all sinks to super-sink with their shelter capacities
  for (const tId of sinks) {
    const node = graph.nodes[tId];
    const cap = node?.capacity || 3000;
    addEdge(tId, superSink, cap, `sink_${tId}`);
  }

  // Add graph network edges with capacities scaled by hazard safety (impassable roads = 0 cap)
  for (const edge of Object.values(graph.edges)) {
    if (edge.isBlocked || edge.hazardRisk >= 0.9) continue;
    // Flow capacity reduced as hazard risk increases
    const effectiveCap = Math.round(edge.capacity * (1 - edge.hazardRisk * 0.7));
    addEdge(edge.source, edge.target, effectiveCap, edge.id);
    addEdge(edge.target, edge.source, effectiveCap, edge.id);
  }

  steps.push({
    stepIndex: 0,
    description: `Constructed residual flow network with Super-Source feeding ${sources.length} hazard-affected zones and Super-Sink collecting into ${sinks.length} evacuation shelters.`,
    highlightedNodeIds: [...sources, ...sinks],
    highlightedEdgeIds: []
  });

  // Dinic BFS Level Graph
  const level: Record<string, number> = {};
  const ptr: Record<string, number> = {};

  function bfsLevelGraph(): boolean {
    for (const id of allNodes) level[id] = -1;
    level[superSource] = 0;
    const queue = [superSource];

    while (queue.length > 0) {
      const v = queue.shift()!;
      for (const edge of adj[v]) {
        if (edge.capacity - edge.flow > 0 && level[edge.to] === -1) {
          level[edge.to] = level[v] + 1;
          queue.push(edge.to);
        }
      }
    }
    return level[superSink] !== -1;
  }

  function dfsBlockingFlow(v: string, pushed: number): number {
    if (pushed === 0 || v === superSink) return pushed;
    for (let cid = ptr[v]; cid < adj[v].length; cid++) {
      ptr[v] = cid;
      const edge = adj[v][cid];
      const tr = edge.to;
      if (level[v] + 1 !== level[tr] || edge.capacity - edge.flow === 0) continue;

      const trPushed = dfsBlockingFlow(tr, Math.min(pushed, edge.capacity - edge.flow));
      if (trPushed === 0) continue;

      edge.flow += trPushed;
      adj[tr][edge.revIndex].flow -= trPushed;
      return trPushed;
    }
    return 0;
  }

  let totalMaxFlow = 0;
  let phase = 0;

  while (bfsLevelGraph()) {
    phase++;
    for (const id of allNodes) ptr[id] = 0;
    let pushed = 0;
    while ((pushed = dfsBlockingFlow(superSource, Infinity)) > 0) {
      totalMaxFlow += pushed;
    }

    if (steps.length < 25) {
      steps.push({
        stepIndex: steps.length,
        description: `Dinic Phase #${phase}: Layered graph augmented. Cumulative evacuation flow reached ${totalMaxFlow.toLocaleString()} evacuees/hour.`,
        highlightedNodeIds: sources.slice(0, 5),
        highlightedEdgeIds: []
      });
    }
  }

  // Compute Min-Cut edges (edges in original graph crossing from reachable in residual graph to unreachable)
  const reachableInResidual = new Set<string>();
  const cutQueue = [superSource];
  reachableInResidual.add(superSource);

  while (cutQueue.length > 0) {
    const u = cutQueue.shift()!;
    for (const edge of adj[u]) {
      if (edge.capacity - edge.flow > 0 && !reachableInResidual.has(edge.to)) {
        reachableInResidual.add(edge.to);
        cutQueue.push(edge.to);
      }
    }
  }

  const minCutEdgeIds = new Set<string>();
  const edgeFlowAttrs: Record<string, { currentFlow: number; isMinCut: boolean }> = {};

  for (const edge of Object.values(graph.edges)) {
    const uReachable = reachableInResidual.has(edge.source);
    const vReachable = reachableInResidual.has(edge.target);

    // If one end is reachable from source in residual graph and other is not, it's a critical bottleneck cut edge
    if ((uReachable && !vReachable) || (!uReachable && vReachable)) {
      minCutEdgeIds.add(edge.id);
      edgeFlowAttrs[edge.id] = { currentFlow: edge.capacity, isMinCut: true };
    }
  }

  steps.push({
    stepIndex: steps.length,
    description: `Max-Flow Min-Cut Theorem applied: Identified ${minCutEdgeIds.size} critical bottleneck road segments throttling evacuation capacity.`,
    highlightedNodeIds: [],
    highlightedEdgeIds: Array.from(minCutEdgeIds)
  });

  const elapsed = performance.now() - startTime;
  return {
    algorithmName: "Dinic's Maximum Evacuation Flow & Min-Cut Bottlenecks",
    executionTimeMs: Math.round(elapsed * 100) / 100,
    summary: `Maximum evacuation throughput: ${totalMaxFlow.toLocaleString()} evacuees/hour. Identified ${minCutEdgeIds.size} saturated Min-Cut bottleneck arteries.`,
    steps,
    edgeAttributes: edgeFlowAttrs,
    customMetrics: {
      'Max Evacuation Throughput': `${totalMaxFlow.toLocaleString()} people/hr`,
      'Bottleneck Cut Roads': minCutEdgeIds.size,
      'Active Danger Sources': sources.length,
      'Available Shelters': sinks.length
    }
  };
}

File: backend/algorithms/max_flow.py
import time
from collections import deque
from typing import Dict, List, Set, Any, Optional
from ..models.graph import GraphData
from ..models.algorithm_results import AlgorithmExecutionResult, AlgorithmStep

class FlowEdge:
    def __init__(self, u: str, v: str, cap: float, orig_edge_id: str):
        self.u = u
        self.v = v
        self.cap = cap
        self.flow = 0.0
        self.orig_edge_id = orig_edge_id
        self.rev_index = -1

def run_dinic_max_evacuation_flow(
    graph: GraphData,
    source_node_ids: Optional[List[str]] = None,
    sink_node_ids: Optional[List[str]] = None
) -> AlgorithmExecutionResult:
    start_time = time.perf_counter()
    steps: List[AlgorithmStep] = []

    sources = source_node_ids if source_node_ids else [
        nid for nid, n in graph.nodes.items() if n.hazardRisk > 0.4 or n.type == 'residential'
    ]
    sinks = sink_node_ids if sink_node_ids else [
        nid for nid, n in graph.nodes.items() if n.type in ('shelter', 'hospital')
    ]

    super_source = '__SUPER_SOURCE__'
    super_sink = '__SUPER_SINK__'
    all_nodes = [super_source, super_sink] + list(graph.nodes.keys())

    adj: Dict[str, List[FlowEdge]] = {nid: [] for nid in all_nodes}

    def add_edge(u: str, v: str, cap: float, edge_id: str):
        fwd = FlowEdge(u, v, cap, edge_id)
        bwd = FlowEdge(v, u, 0.0, edge_id)
        fwd.rev_index = len(adj[v])
        bwd.rev_index = len(adj[u])
        adj[u].append(fwd)
        adj[v].append(bwd)

    # Super-source to risk nodes
    for s_id in sources:
        node = graph.nodes.get(s_id)
        pop = max(node.population, 500) if node else 1000
        add_edge(super_source, s_id, float(pop), f"src_{s_id}")

    # Sinks to super-sink
    for t_id in sinks:
        node = graph.nodes.get(t_id)
        cap = float(node.capacity) if node and node.capacity else 3000.0
        add_edge(t_id, super_sink, cap, f"sink_{t_id}")

    # Graph network edges
    for edge in graph.edges.values():
        if edge.isBlocked or edge.hazardRisk >= 0.9:
            continue
        eff_cap = round(edge.capacity * (1.0 - edge.hazardRisk * 0.7))
        add_edge(edge.source, edge.target, float(eff_cap), edge.id)
        add_edge(edge.target, edge.source, float(eff_cap), edge.id)

    steps.append(AlgorithmStep(
        stepIndex=0,
        description=f"Residual network constructed with Super-Source feeding {len(sources)} hazard zones and Super-Sink draining to {len(sinks)} evacuation sanctuaries.",
        highlightedNodeIds=sources[:4] + sinks[:4],
        highlightedEdgeIds=[]
    ))

    level: Dict[str, int] = {}
    ptr: Dict[str, int] = {}

    def bfs_level() -> bool:
        for nid in all_nodes:
            level[nid] = -1
        level[super_source] = 0
        q = deque([super_source])
        while q:
            u = q.popleft()
            for edge in adj[u]:
                if edge.cap - edge.flow > 0 and level[edge.v] == -1:
                    level[edge.v] = level[u] + 1
                    q.append(edge.v)
        return level[super_sink] != -1

    def dfs_block(u: str, pushed: float) -> float:
        if pushed == 0 or u == super_sink:
            return pushed
        for cid in range(ptr[u], len(adj[u])):
            ptr[u] = cid
            edge = adj[u][cid]
            tr = edge.v
            if level[u] + 1 != level[tr] or edge.cap - edge.flow == 0:
                continue
            tr_pushed = dfs_block(tr, min(pushed, edge.cap - edge.flow))
            if tr_pushed == 0:
                continue
            edge.flow += tr_pushed
            adj[tr][edge.rev_index].flow -= tr_pushed
            return tr_pushed
        return 0.0

    total_max_flow = 0.0
    phase = 0
    while bfs_level():
        phase += 1
        for nid in all_nodes:
            ptr[nid] = 0
        while True:
            pushed = dfs_block(super_source, float('inf'))
            if pushed <= 0:
                break
            total_max_flow += pushed

        if len(steps) < 25:
            steps.append(AlgorithmStep(
                stepIndex=len(steps),
                description=f"Dinic Phase #{phase}: Level graph augmented. Evacuation throughput reached {int(total_max_flow):,} people/hr.",
                highlightedNodeIds=sources[:4],
                highlightedEdgeIds=[]
            ))

    # Min-Cut detection
    reachable_in_residual = set()
    cut_q = deque([super_source])
    reachable_in_residual.add(super_source)

    while cut_q:
        u = cut_q.popleft()
        for edge in adj[u]:
            if edge.cap - edge.flow > 0 and edge.v not in reachable_in_residual:
                reachable_in_residual.add(edge.v)
                cut_q.append(edge.v)

    min_cut_edges: Set[str] = set()
    edge_flow_attrs: Dict[str, Dict[str, Any]] = {}

    for edge in graph.edges.values():
        u_reach = edge.source in reachable_in_residual
        v_reach = edge.target in reachable_in_residual
        if (u_reach and not v_reach) or (not u_reach and v_reach):
            min_cut_edges.add(edge.id)
            edge_flow_attrs[edge.id] = {'currentFlow': edge.capacity, 'isMinCut': True}

    steps.append(AlgorithmStep(
        stepIndex=len(steps),
        description=f"Max-Flow Min-Cut Theorem applied: Identified {len(min_cut_edges)} saturated bottleneck roads throttling evacuation throughput.",
        highlightedNodeIds=[],
        highlightedEdgeIds=list(min_cut_edges)
    ))

    elapsed = (time.perf_counter() - start_time) * 1000

    return AlgorithmExecutionResult(
        algorithmName="Dinic's Maximum Evacuation Flow & Min-Cut Bottlenecks (Python)",
        executionTimeMs=round(elapsed, 2),
        summary=f"Max Evacuation Throughput: {int(total_max_flow):,} evacuees/hour. Saturated Min-Cut Bottlenecks: {len(min_cut_edges)} arterial segments.",
        steps=steps,
        edgeAttributes=edge_flow_attrs,
        customMetrics={
            'Max Evacuation Throughput': f"{int(total_max_flow):,} people/hr",
            'Bottleneck Cut Roads': len(min_cut_edges),
            'Active Hazard Sources': len(sources),
            'Designated Shelters': len(sinks),
            'Execution Speed': f"{round(elapsed, 2)} ms"
        }
    )
